### ISSUE TITLE
`@dropbox/sign` module resolution error

### ISSUE THREAD

**Issue by reefloretto** [hellosign/hellosign-openapi#327]
I'm trying to integrate the Dropbox Sign Node SDK in my application. I'm building a Next.js application and am using Next 14 with Node 20.8.1. I'm also using Yarn workspaces (for what that's worth). When I try to compile any code that imports from the `@dropbox/sign` package, I get the following Node module resolution errors:

```
Failed to compile.

../node_modules/@dropbox/sign/dist/api.js
Module not found: Can't resolve './aes'

https://nextjs.org/docs/messages/module-not-found

Import trace for requested module:
./app/admin/dropbox/page.tsx

../node_modules/@dropbox/sign/dist/api.js
Module not found: Can't resolve './blowfish'

https://nextjs.org/docs/messages/module-not-found

Import trace for requested module:
./app/admin/dropbox/page.tsx

../node_modules/@dropbox/sign/dist/api.js
Module not found: Can't resolve './cipher-core'

https://nextjs.org/docs/messages/module-not-found

Import trace for requested module:
./app/admin/dropbox/page.tsx

../node_modules/@dropbox/sign/dist/api.js
Module not found: Can't resolve './core'

https://nextjs.org/docs/messages/module-not-found

Import trace for requested module:
./app/admin/dropbox/page.tsx

../node_modules/@dropbox/sign/dist/api.js
Module not found: Can't resolve './enc-base64'

https://nextjs.org/docs/messages/module-not-found

Import trace for requested module:
./app/admin/dropbox/page.tsx

> Build failed because of webpack errors
```

These all look like relative imports, which may understandably be raising because, for example, the `crypto-js` package may not expose its modules as siblings of the `../node_modules/@dropbox/sign/dist/api.js` file which appears to be executing.

I also [don't think I'm the only one experiencing this issue](https://www.dropboxforum.com/t5/Dropbox-API-Support-Feedback/Can-t-resolve-aes-error-when-using-dropbox-sign/td-p/721112).

Can anyone provide any guidance on this?

Thank you.

**Comment by jtreminio-dropbox** [comment]
Howdy! Do you have an example repo I can clone that encounters this problem?

**Comment by jtreminio-dropbox** [comment]
This might be fixed by changing https://github.com/hellosign/hellosign-openapi/blob/3becc8a4b09237470642201b0ab40c8fce6aa34a/sdks/node/model/eventCallbackHelper.ts#L2 to

```typescript
import HmacSHA256 from 'crypto-js/hmac-sha256';
```

I'll try spinning up a nextjs hello-world app and testing this otu.

**Comment by jtreminio-dropbox** [comment]
Howdy again.

I have created this simple Hello World repo for using `@dropbox/sign` with Next.js: https://github.com/hellosign/nextjs-blog

The trick is [next.config.js](https://github.com/hellosign/nextjs-blog/blob/main/next.config.js) must ignore `fs` and `axios` on client-side rendering.

Make sure to invoke the SDK package only within [getStaticProps](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-static-props) or [getServerSideProps](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-server-side-props) functions.

**Comment by reefloretto** [comment]
hey @jtreminio-dropbox -- thanks for taking a look at this! i'm going to try the `next.config.js` trick right now

**Comment by reefloretto** [comment]
hey @jtreminio-dropbox!

i've created [this repo](https://github.com/reefloretto/dropbox-sign-repro) to hopefully provide more assistance in reproducing the build error. some additional context that might be helpful--i'm using next.js's "[app router](https://nextjs.org/docs/app)" which means i'm not using `getStaticProps` or `getServerSideProps` in my application, in favor of alternative methods of executing code on the server.

i hope this helps, and please let me know if you'd like any additional context! (i'd love to get to the bottom of this 😅)

**Comment by jtreminio-dropbox** [comment]
Sorry, I don't know enough about next.js or the app router feature to really be able to assist further. However, as demonstrated in my repo example, the problem is not caused by how the 3rd-party modules are being referenced from within `dist/api.js`, but that the SDK will not work with client-side code, it has to be server-side.

I don't know how next.js's app router handles this.

**Comment by reefloretto** [comment]
all good, @jtreminio-dropbox thanks for taking a look!

**Comment by heraldofsolace** [comment]
I'm running into this issue too when calling the SDK from an API route

**Comment by raphaellopes** [comment]
Hello, I am experiencing this issue on the 1.3.0 version when I run it using `npm run build`. I am not using nextjs, my api is using fastify.

**Comment by jtreminio-dropbox** [comment]
I don't know how the follow up comments didn't get to me, but I've seen this issue is still causing problems, rest assured we are looking into this.

**Comment by raphaellopes** [comment]
> I don't know how the follow up comments didn't get to me, but I've seen this issue is still causing problems, rest assured we are looking into this.

Hello @jtreminio-dropbox . I was doing some local test, and I noticed that the `crypto-js` lib is been used only in this file:
https://github.com/hellosign/hellosign-openapi/blob/93ecfa52116bebb0d6cc6f1254d42426af1fdc4d/sdks/node/model/eventCallbackHelper.ts#L1-L22

After change the lib `crypto-js` to the native node `crypto`, the build seems to work for me. I haven't tested it yet to see if my app still works as expected. Here are the changes I've made:
```ts
import * as crypto from "crypto";
...
const hmac = crypto.createHmac("sha256", apiKey);
const hash = hmac.update(
  `${eventCallback.event.eventTime}${eventCallback.event.eventType}`,
);
```

**Comment by jtreminio-dropbox** [comment]
Thanks @raphaellopes - I'll test this change out and see if it works as expected

**Comment by jtreminio-dropbox** [comment]
@reefloretto @eifo @heraldofsolace Most likely you've moved on due to my lack of response, but if you are able to I'd appreciate testing out the SDK with the change to native crypto library. I've made the package available at https://github.com/hellosign/hellosign-openapi/pull/360#issuecomment-2015414694

I'll wait until next week to merge, barring any bug reports. We'll also test on our side as well.

**Comment by eifo** [comment]
I'm currently out of office for the following weeks and will not be able to
test but looking forward for this solution. Please let know if matter is
resolved.

**Comment by jtreminio-dropbox** [comment]
Fixed by #360
